### PR TITLE
Set the TFM value when the TFM is changed from the property page.

### PR DESCRIPTION
**Pull request by srivatsn**
When the TFM is changed from the application page, the default behavior is to just set the property. However two things needs to happen instead - the TFM needs to be broken down into its constituent parts (TFIdentfier, TFVersion, TFProfile) and those need to get set. Also, we need to call the IVsRetargetingService to set the value which causes a project reload so that some components can flush caches. Both of these are implemented by CPS if set the property through the hierarchy. 

Also noticed that the interceptor for the TargetFrameWork property doesn't depend on anything in the VS layer anymore (it used to set through the hierarchy) - so I've moved that down to the managed layer.

@dotnet/project-system

### CODE REVIEW

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/TargetFrameworkMonikerValueProvider.cs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        }
+
+        public override Task<string> OnSetPropertyValueAsync(string unevaluatedPropertyValue, IProjectProperties defaultProperties, IReadOnlyDictionary<string, string> dimensionalConditions = null)
+        {
+            // CPS implements IVsHierarchy.SetProperty for the TFM property to call through the multi-targeting service and change the TFM.
+            // This causes the project to be reloaded after changing the values.
+            // Since the property providers are called under a write-lock, trying to reload the project on the same context fails saying it can't load the project
+            // if a lock is held. We are not going to write to the file under this lock (we return null from this method) and so we fork execution here to schedule
+            // a lambda on the UI thread and we don't pass the lock information from this context to the new one. 
+            _unconfiguredProjectVsServices.ThreadingService.Fork(() =>
```
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/TargetFrameworkMonikerValueProvider.cs]: Would like some scrutiny of this. This is what I've found works through trial and error but not sure if there any other threading implications of this call.

Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/TargetFrameworkMonikerValueProvider.cs]: I remember you seeming more convinced offline at standup last week. Are you ok with this now?

Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/TargetFrameworkMonikerValueProvider.cs]: I'm fine with this. @davkean do you have any concerns?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/ConfigurationGeneral.xaml [hunk 2]
```diff
@@ -20,10 +20,26 @@
             <DataSource Persistence="ImplicitProjectFile" Label="Globals" HasConfigurationCondition="False" />
         </StringProperty.DataSource>
     </StringProperty>
-    <DynamicEnumProperty Name="TargetFrameworkMoniker" DisplayName="Target Framework Moniker" EnumProvider="TargetFrameworkMoniker" />
-    <StringProperty Name="TargetFrameworkIdentifier" Visible="False" />
-    <StringProperty Name="TargetFrameworkVersion" Visible="False" />
-    <StringProperty Name="TargetFrameworkProfile" Visible="False" />
+    <StringProperty Name="TargetFrameworkIdentifier" DisplayName="Target Framework Identifier">
```
Comment (basoundr) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/ConfigurationGeneral.xaml]: What is the implication of changing the property to have the extra attributes?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/TargetFrameworkMonikerValueProvider.cs (hunk 1 above)
Comment (basoundr) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/TargetFrameworkMonikerValueProvider.cs]: What is the guarantee that this lambda is going to be executed after the write-lock is relinquished? Does CPS guarantee anything here?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/ConfigurationGeneral.xaml (hunk 2 above)
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/ConfigurationGeneral.xaml]: The HasConfigurationCondition="False" specifices that this value is not specific to configuration. So when writing out the value in a project file, it won't put it under a specific configuration.

**Comment by srivatsn** [comment]
@davkean can you take a look?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/TargetFrameworkMonikerValueProvider.cs (hunk 1 above)
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/TargetFrameworkMonikerValueProvider.cs]: @basoundr That's what HideLocks is doing. This looks sane.

**Comment by davkean** [comment]
:+1:

**Pull request merged**